In [71]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import seaborn as sns
import numpy as np

from sklearn.linear_model import LinearRegression
from sklearn.svm import SVR
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error

#https://www.kaggle.com/c/boston-housing/overview

df = pd.read_csv('homework.csv')
df

ID     crim    zn  indus  chas    nox     rm   age     dis  rad  tax  \
0      1  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1      2  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2      4  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
3      5  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
4      7  0.08829  12.5   7.87     0  0.524  6.012  66.6  5.5605    5  311   
..   ...      ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
328  500  0.17783   0.0   9.69     0  0.585  5.569  73.5  2.3999    6  391   
329  502  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
330  503  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
331  504  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
332  506  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     ptratio   black  lstat  medv  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14  21.6  
2       18.7  394.63   2.94  33.4  
3       18.7  396.90   5.33  36.2  
4       15.2  395.60  12.43  22.9  
..       ...     ...    ...   ...  
328     19.2  395.77  15.10  17.5  
329     21.0  391.99   9.67  22.4  
330     21.0  396.90   9.08  20.6  
331     21.0  396.90   5.64  23.9  
332     21.0  396.90   7.88  11.9  

[333 rows x 15 columns]

In [5]:
#проверяем на NA
df.info()
#NA отсутствуют

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 333 entries, 0 to 332
Data columns (total 15 columns):
ID         333 non-null int64
crim       333 non-null float64
zn         333 non-null float64
indus      333 non-null float64
chas       333 non-null int64
nox        333 non-null float64
rm         333 non-null float64
age        333 non-null float64
dis        333 non-null float64
rad        333 non-null int64
tax        333 non-null int64
ptratio    333 non-null float64
black      333 non-null float64
lstat      333 non-null float64
medv       333 non-null float64
dtypes: float64(11), int64(4)
memory usage: 39.1 KB


In [6]:
#посмотрим на данные с точки зрения статистики
df.describe()

ID        crim          zn       indus        chas         nox  \
count  333.000000  333.000000  333.000000  333.000000  333.000000  333.000000   
mean   250.951952    3.360341   10.689189   11.293483    0.060060    0.557144   
std    147.859438    7.352272   22.674762    6.998123    0.237956    0.114955   
min      1.000000    0.006320    0.000000    0.740000    0.000000    0.385000   
25%    123.000000    0.078960    0.000000    5.130000    0.000000    0.453000   
50%    244.000000    0.261690    0.000000    9.900000    0.000000    0.538000   
75%    377.000000    3.678220   12.500000   18.100000    0.000000    0.631000   
max    506.000000   73.534100  100.000000   27.740000    1.000000    0.871000   

               rm         age         dis         rad         tax     ptratio  \
count  333.000000  333.000000  333.000000  333.000000  333.000000  333.000000   
mean     6.265619   68.226426    3.709934    9.633634  409.279279   18.448048   
std      0.703952   28.133344    1.981123    8.742174  170.841988    2.151821   
min      3.561000    6.000000    1.129600    1.000000  188.000000   12.600000   
25%      5.884000   45.400000    2.122400    4.000000  279.000000   17.400000   
50%      6.202000   76.700000    3.092300    5.000000  330.000000   19.000000   
75%      6.595000   93.800000    5.116700   24.000000  666.000000   20.200000   
max      8.725000  100.000000   10.710300   24.000000  711.000000   21.200000   

            black       lstat        medv  
count  333.000000  333.000000  333.000000  
mean   359.466096   12.515435   22.768769  
std     86.584567    7.067781    9.173468  
min      3.500000    1.730000    5.000000  
25%    376.730000    7.180000   17.400000  
50%    392.050000   10.970000   21.600000  
75%    396.240000   16.420000   25.000000  
max    396.900000   37.970000   50.000000

In [127]:
#делим выборку на тестовую и тренировочную выборки
x = df[['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax', 'ptratio', 'black', 'lstat']]
y = df[['medv']]
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3)

In [200]:
#делаем "стандартизацию" данных и записываем в формат dataFrame
#stdsc = StandardScaler()
#x_train = stdsc.fit_transform(x_train)
#x_test = stdsc.transform(x_test)
#x_train = pd.DataFrame(x_train, columns = x.columns)
#x_test = pd.DataFrame(x_test, columns = x.columns)

In [128]:
#Создадим модель линейной регрессии, проведем тренировку на тренировочных данных и вычислим соответствующие коэффициенты для признаков
lr = LinearRegression()
lr.fit(x_train, y_train)
print(lr.coef_, lr.intercept_)
y_predict = lr.predict(x_test)
np.transpose(y_predict)

[[-1.18913970e-02  4.96874747e-02  6.77053398e-02  5.83989143e+00
  -1.61575895e+01  4.29739971e+00 -8.13172905e-03 -1.56836368e+00
   2.56020038e-01 -1.11588580e-02 -7.91876678e-01  1.17885975e-02
  -5.09110730e-01]] [28.43923191]


array([[25.17581586, 23.7870287 , 25.1759959 , 24.41943056, 28.12126045,
        10.61533537, 24.14957378, 20.44647126, 19.34405735, 17.99200759,
        34.12936357, 23.34999767,  4.37956284, 20.20458178, 37.10387818,
        14.35997152, 24.24327914, 17.79679304, 42.87830886, 26.35948375,
        26.04606847, 20.32992542, 33.27583478, 19.50023725, 19.56088945,
        23.1653997 , 28.15205546, 38.99106766, 27.17190747, 15.01818775,
        32.91043028, 38.8335698 , 31.07218588, 19.08241355, 16.57926221,
        31.54872286, 21.01061515, 21.65928977, 34.36092994, 19.68039496,
        13.34087044, 20.48978363, 14.12438301, 22.78636156, 23.78796478,
        24.70383154, 19.14741725, 30.03804192, 26.38273293, 24.01436054,
        16.11275491,  6.8615222 , 23.05971844, 23.14668668, 25.41404481,
        35.8017945 , 30.27728919, 11.13998585, 35.43728954, 14.68662568,
        23.74079197, 12.39472955, 24.18822227, 38.00719547, 30.69659243,
        33.41060666, 24.72295279, 19.4442876 , 21.0

In [184]:
lr_mse = mean_squared_error(y_test, y_predict)
lr_mae = mean_absolute_error(y_test, y_predict)
lr_r2 = r2_score(y_test, y_predict)
print('Mean Square error: ',lr_mse, '\nMean Absolute error: ', lr_mae, '\nR^2: ', lr_r2)

Mean Square error:  25.37978275754114 
Mean Absolute error:  3.4243492868670966 
R^2:  0.7273145726709753


In [160]:
#проверка
len_0 = x_test.iloc[[0]].shape[1]
x_arr = list(x_test.iloc[[0]].values[0])
y_arr = list(np.transpose(lr.coef_[0]))
print('Расчетное значение через коэффициенты: ', sum([x_arr[i]*y_arr[i] for i in range(len_0)]) + lr.intercept_[0])
print('Расчетное значение через функцию predict: ', y_predict[0][0])
print('Реальное значение: ', list(y_test.iloc[[0]].values[0])[0])

Расчетное значение через коэффициенты:  25.17581586322584
Расчетное значение через функцию predict:  25.17581586322584
Реальное значение:  28.1


In [187]:
#Создадим модель линейной регрессии SVR
svr = SVR(gamma = 'scale', C=1.0, epsilon=0.2)
svr.fit(np.array(x_train), np.ravel(np.array(y_train)))


SVR(C=1.0, cache_size=200, coef0=0.0, degree=3, epsilon=0.2, gamma='scale',
    kernel='rbf', max_iter=-1, shrinking=True, tol=0.001, verbose=False)

In [188]:
y_pred_svr = svr.predict(x_test)
y_pred_svr

array([23.4538023 , 23.14108847, 23.73397134, 23.32699026, 23.70782103,
       17.36056494, 23.23224561, 20.32055011, 17.36963375, 20.84609254,
       22.46321745, 17.18227963, 15.97650294, 22.69412999, 22.65008607,
       17.31938028, 23.60091797, 22.50117859, 24.75200579, 23.35815169,
       22.52362048, 21.07598686, 24.38648248, 22.95775007, 17.5430228 ,
       17.24171756, 24.13359553, 24.163107  , 22.34458931, 23.56779289,
       24.3821089 , 24.16258883, 24.46458619, 13.91534366, 14.05548021,
       24.29510129, 20.66256149, 24.11728097, 24.46104567, 17.72026895,
       16.99625477, 22.58325064, 17.3498625 , 22.62460232, 17.12890747,
       23.14568128, 17.30330356, 21.9362443 , 20.64235777, 23.19269275,
       23.82835528, 17.30158119, 17.45046315, 22.76724148, 22.9496631 ,
       24.33725856, 23.19615806, 22.10987706, 20.3312133 , 17.37085791,
       17.16101585, 15.41410921, 24.28231031, 22.7008506 , 22.88964811,
       24.53346476, 23.04747843, 15.68125256, 24.10690304, 19.16

In [198]:
svr_mse = mean_squared_error(y_test, y_pred_svr)
svr_mae = mean_absolute_error(y_test, y_pred_svr)
svr_r2 = r2_score(y_test, y_pred_svr)
print('Mean Square error: ',svr_mse, '\nMean Absolute error: ', svr_mae, '\nR^2: ', svr_r2)

Mean Square error:  75.41362425688999 
Mean Absolute error:  5.881274848508453 
R^2:  0.1897410409941248


In [190]:
#Создадим модель DecisionTreeRegressor
dtr = DecisionTreeRegressor()
dtr.fit(np.array(x_train), np.ravel(np.array(y_train)))

DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
                      max_leaf_nodes=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      presort=False, random_state=None, splitter='best')

In [195]:
y_pred_dtr = dtr.predict(x_test)
y_pred_dtr

array([21.4, 25.2, 22.6, 25. , 27.9,  8.3, 23.9, 17.7, 15.4, 13.8, 25.1,
       27.9,  5. , 22.9, 44.8,  8.8, 22. , 12.8, 50. , 27.1, 12.8, 20.8,
       29.1, 16.6, 20.2, 22.6, 22. , 31.5, 23.6, 17.3, 33.4, 37.6, 29.1,
       12.5, 23.1, 24.5, 17.4, 23.1, 30.5, 20.6, 22.6, 26.4, 12.3, 19.3,
       50. , 22.6, 15.4, 29.4, 26.5, 21.4, 20.5, 13.8, 18.6, 16.1, 22.6,
       39.8, 44.8, 12.8, 24.3, 13.6, 15.4,  8.3, 25.3, 50. , 32.5, 33.4,
       23.3, 16.1, 23.1, 17.4, 22. , 17.4, 23.6, 20.2, 12.8, 17.5, 14.5,
       50. , 22.3, 50. , 17.7, 34.9,  5. , 25. , 24.7, 22.5, 22.3, 23.1,
       13.8, 18. , 16.1, 18.3, 16.2, 33.4, 13.8, 19.2, 25. ,  7.2, 20.5,
       24.5])

In [199]:
dtr_mse = mean_squared_error(y_test, y_pred_dtr)
dtr_mae = mean_absolute_error(y_test, y_pred_dtr)
dtr_r2 = r2_score(y_test, y_pred_dtr)
print('Mean Square error: ',dtr_mse, '\nMean Absolute error: ', dtr_mae, '\nR^2: ', dtr_r2)

Mean Square error:  32.118 
Mean Absolute error:  3.5060000000000007 
R^2:  0.6549178281539347


Вывод: все модели показали результат не очень точный, где коэффициент детерминации r2 0.72 был у модели линейной регресии (получается что, лучший метода), то есть модель на 72% кореллирует с исходными данными. хуже результат был у модели дерева решений 0.65 и совсем плохой 0,18. MAE и MSE лучше всех также  у линейной регрессии (менбшее значение, значит меньшая ошибка) и затем у дерева решений у дерева решений. Хуже у SVR